### PR TITLE
Display different pages for completed cases and inactive iac

### PR DESCRIPTION
**Pull request by LukeLoze1**
Display different pages for completed cases and inactive iac codes for other reasons

# Motivation and Context
Currently this app displays a 'case completed' page when any inactive iac is used.  This is correct if the case is completed.  However it's possible in social for a refusal 562 for an iac to be set to inactive.  In this case the page should be say that the iac is inactive

# What has changed

- New html page inactive-iac.html added.  
- New code to check if an inactive iac is for a completed case or not.

- Renamed some variables and functions.  Extracted sample unit type check to seperate

- test_post_index_iac_active_missing, test_post_index_case_complete,
 test_missing_case_group_status, test_post_index_iac_inactive

# How to test?

    Checkout this branch.  
    In ras-rm-docker-dev do **make pull**
    In respondent-home-ui run docker **build . -t sdcplatform/respondent-home-ui**
    In ras-rm-docker-dev run **make up** and wait for system to be fully up.
    
- **Test Locally**
   - In respondent-home-ui run **make install** then run **make test** all tests should pass
   - Get latest copy of acceptance tests and run **make acceptance_tests** all should pass
 
- **Manually Test**
   - Browse go to http://localhost:8003/survey and create a new LMS social Collection exercise to 
      start in the next few minutes.  Use the attached DUMMYSAMPLE.csv (given suffix .txt to upload 
      here, but remove when downloaded
   - Browse to http://localhost:8086/ and enter **np10 8xg**  as the postcode, choose a case and 
      create a new iac for it (record this)
   - Browse to http://localhost:9092/ and use this code, the browser should attempt (and fail) to 
      redirect to an EQ, page not found etc
   - Mock the affect of a normally completed case by using pgAdmin change iac.iac active to false for 
      the new iac you have generated (your new one should be at the bottom).  On the table 
      casesvc.casegroup update the case  (sampleunitref should match, eg DUM100012).  Then reuse 
      the iac code in a refreshed http://localhost:9092/ and it should now display a page stating 'Study 
      Completed'.
   - Now in http://localhost:8086/ choose a new case and create a new iac for it, record this.  Then 
      change the status on this case to 562 with the UI.  Now browse to http://localhost:9092/ and 
      use the iac.  This should now display a page with 'Access code disabled'

#What I did locally
- I used a standard local environment and ran manual test as above.  
- Ran **make test** to test project
- Ran **make acceptance_tests** to test the whole system

[DUMMYSample.csv.txt](https://github.com/ONSdigital/respondent-home-ui/files/2671743/DUMMYSample.csv.txt)

No new automated tests as this only touches one service and doesn't break any existing tests.

# Links
https://trello.com/c/KKYcX9A8/331-sus057-manage-error-messages-for-disabled-completed-closed-surveys

# Screenshots (if appropriate):

### CODE REVIEW

File: app/templates/inactive-iac.html [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{% extends "base.html" %}
+
+{% block title %}Access Code Inactive{% endblock title %}
+
+{% block content %}
+
+<h1 class="saturn" data-ga="error" data-ga-category="info" data-ga-action="info-message" data-ga-label="info_type = CODE_USED">
```
Comment (jcox-dev) [inline app/templates/inactive-iac.html]:
```suggestion
<h1 class="saturn" data-ga="error" data-ga-category="info" data-ga-action="info-message" data-ga-label="info_type = CODE_INACTIVE">
```

File: app/handlers.py [hunk 2]
```diff
@@ -56,13 +57,39 @@ def join_iac(data, expected_length=12):
...
+                    raise CompletedCaseError
+            except KeyError:
+                logger.warn("Field case_json['caseGroup']['caseGroupStatus'] not found")
+
+            logger.info('Attempt to use inactive iac for incomplete case')
+            raise InactiveIACError
+
+    def check_case_sample_unit_type_valid(self, case_json):
+        try:
+            assert case_json['sampleUnitType'] == 'H'
```
Comment (AdamHawtin) [inline app/handlers.py]: Best not to use asserts, under certain options the interpreter can ignore the statement completely.

Comment (jcox-dev) [reply app/handlers.py]: @AdamHawtin this already existed prior to these changes and is a valid use case for this (should only be skipped if certain optimisation flags were used at runtime) 👍

File: app/handlers.py [hunk 3]
```diff
@@ -56,13 +57,39 @@ def join_iac(data, expected_length=12):
         return combined
 
     @staticmethod
-    def validate_case(case_json):
-        if not case_json.get("active", False):
-            raise InactiveCaseError
+    def validate_iac_active(iac_json, case_json):
+        if not iac_json.get("active", False):
+            try:
+                if case_json['caseGroup']['caseGroupStatus'] in ['COMPLETE', 'COMPLETEDBYPHONE']:
```
Comment (liamtoozer) [inline app/handlers.py]: Minor (pedantic!) comment but I don't think response-ops-social-ui currently supports change of status to `COMPLETEDBYPHONE`? No harm in keeping this in for future though?

**Comment by rjweeks** [comment]
[SUS057: Manage error messages for disabled / completed / closed surveys](https://trello.com/c/KKYcX9A8/331-sus057-manage-error-messages-for-disabled-completed-closed-surveys)

File: app/error_handlers.py [hunk 4]
```diff
@@ -44,11 +45,16 @@ def create_error_middleware(overrides):
     return middleware_handler
 
 
-async def inactive_case(request):
-    logger.info("Attempt to use an inactive access code")
+async def completed_case(request):
+    logger.info("Attempt to use an inactive iac for a completed case")
```
[the remark this answers is not in the document]
Comment (LukeLoze1) [reply app/error_handlers.py]: I believe it's easier to access this information (collex id) upstream of here in the handlers where these Exceptions are thrown.

[the remark this answers is not in the document]
Comment (LukeLoze1) [reply app/error_handlers.py]: I've put the collex id in the logging now

File: app/error_handlers.py [hunk 5]
```diff
@@ -44,11 +45,16 @@ def create_error_middleware(overrides):
...
     return aiohttp_jinja2.render_template("completed.html", request, {})
 
 
+async def inactive_iac(request):
+    logger.info("Attempt to use an iac code that is inactive, malformed or iac_details missing active field")
```
[the remark this answers is not in the document]
Comment (LukeLoze1) [reply app/error_handlers.py]: I've put the collex id in the logging now

File: app/handlers.py [hunk 6]
```diff
@@ -56,13 +57,39 @@ def join_iac(data, expected_length=12):
...
+                    logger.info('Attempt to use inactive iac for completed case')
+                    raise CompletedCaseError
+            except KeyError:
+                logger.warn("Field case_json['caseGroup']['caseGroupStatus'] not found")
```
[the remark this answers is not in the document]
Comment (LukeLoze1) [reply app/handlers.py]: collex id now recorded in next log message

**Comment by LukeLoze1** [comment]
@jcox-dev can you re review

**Comment by LukeLoze1** [comment]
@AdamHawtin are you ok with the assert now? after Jamie's comment

**Pull request merged**